pkg/splunk/enterprise/secrets.go

```diff
@@ -0,0 +1,293 @@
+
+	// Check if there is atleast one versionedSecretIdentifier based secret
+	if existingLatestVersion == -1 {
+		// No secret based on versionedSecretIdentifier, create one with version v1
+		latestVersionedSecret, err = ApplySplunkSecret(c, cr, splunkReadableData, GetVersionedSecretName(versionedSecretIdentifier, firstVersion), namespace)
+	} else {
+		// Check if contents of latest versionedSecretIdentifier based secret is different from that of namespace scoped secrets object
+		if !reflect.DeepEqual(splunkReadableData, existingLatestVersionedSecret.Data) {
+			// Different, create a newer version versionedSecretIdentifier based secret
+			latestVersionedSecret, err = ApplySplunkSecret(c, cr, splunkReadableData, GetVersionedSecretName(versionedSecretIdentifier, strconv.Itoa(existingLatestVersion+1)), namespace)
+			return latestVersionedSecret, err
+		}
+		// Latest versionedSecretIdentifier based secret is the the existing latest versionedSecretIdentifier based secret
+		latestVersionedSecret = existingLatestVersionedSecret
+	}
+
+	return latestVersionedSecret, nil
+}
+
+// GetSplunkReadableNamespaceScopedSecretData retreives the namespace scoped secret's data and converts it into Splunk readable format if possible
+func GetSplunkReadableNamespaceScopedSecretData(c splcommon.ControllerClient, namespace string) (map[string][]byte, error) {
+	// Get namespace scoped secret
+	namespaceScopedSecret, err := GetNamespaceScopedSecret(c, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, tokenType := range GetSplunkSecretTokenTypes() {
+		if _, ok := namespaceScopedSecret.Data[tokenType]; !ok {
+			return nil, fmt.Errorf(missingTokenError, tokenType)
+		}
+	}
+
+	// Create data
+	splunkReadableData := make(map[string][]byte)
+
+	// Create individual token type data
+	for _, tokenType := range GetSplunkSecretTokenTypes() {
+		splunkReadableData[tokenType] = namespaceScopedSecret.Data[tokenType]
+	}
+
+	// Create default.yml
+	splunkReadableData["default.yml"] = []byte(fmt.Sprintf(`
+splunk:
+    hec_disabled: 0
+    hec_enableSSL: 0
+    hec_token: "%s"
+    password: "%s"
+    pass4SymmKey: "%s"
+    idxc:
+        secret: "%s"
+    shc:
+        secret: "%s"
+`,
+		namespaceScopedSecret.Data["hec_token"],
+		namespaceScopedSecret.Data["password"],
+		namespaceScopedSecret.Data["pass4SymmKey"],
+		namespaceScopedSecret.Data["idxc_secret"],
+		namespaceScopedSecret.Data["shc_secret"]))
+
+	return splunkReadableData, nil
+}
+
+// ApplySplunkSecret creates/updates a secret using secretData(which HAS to be of ansible readable format) or namespace scoped secret data if not specified
+func ApplySplunkSecret(c splcommon.ControllerClient, cr splcommon.MetaObject, secretData map[string][]byte, secretName string, namespace string) (*corev1.Secret, error) {
+	var current corev1.Secret
+	var newSecretData map[string][]byte
+	var err error
+
+	// Prepare secret data
+	if secretData != nil {
+		newSecretData = secretData
+	} else {
```

Comment:
Curious, when does this happen?

Reply:
ApplySplunkSecret can be used by any application to create/update a secret(with name <secretName>) whose data will be derived from the namespace scoped secret if secretData is passed as nil. It makes is easy when someone wants to use the namespace scoped secret's data, transform it to ansible readable format and create/update secret based on the same. The intention is to make the caller agnostic of secret related operations.